ROX-27622: Move SecurityPolicy CRD to template dir in Helm chart

kylape opened the pull request:
To account for the move, the following additional steps were taken:

* Added a new flag to disable CRD installation: `allowUnsupportedMultipleCentralsUsingHelm=documentationWasReadAndUnderstood`
* Documented how to handle multi-tenancy with the CRD in the templates folder
* Set the flag to disable CRD installation in the operator

Note that upgrading Central on a cluster with multiple installations of Central will likely result in a failure due to CRD management issues described in the readme.

There was an attempt to use a Helm pre-upgrade hook to automate patching the CRD to avoid potential upgrade failures, but unfortunately the pre-upgrade hook is executed after Helm raises the validation failure for the lack of CRD label and annotations.

porridge commented on image/templates/helm/stackrox-central/values.yaml.htpl:
```diff
@@ -552,6 +552,28 @@ central:
+##     management issues are considered unsupported!
+##
+## To enable the option, it needs to be set to the value "documentationWasReadAndUnderstood".
+#allowUnsupportedMultipleCentralsUsingHelm: false
```
Minor nitpick: we say this is a string in config-shape.yaml, but use a bool here. This is a minor cognitive discomfort. Perhaps `"no"`?

porridge commented on image/templates/helm/stackrox-central/values.yaml.htpl:
```diff
@@ -552,6 +552,28 @@ central:
+##     introducing the potential for versioning mismatches.
+##  2. Uninstallation of the first installation will remove the CRD, thus removing all custom
+##     `SecurityPolicy` resources from ALL Central installations.
+##  3. Downdgrading of the first installation could remove the CRD as well.
```
```suggestion
##  3. Downgrading of the first installation could remove the CRD as well.
```

porridge commented on image/templates/helm/stackrox-central/README.md.htpl:
```diff
@@ -218,3 +218,26 @@ Note that because Scanner V4 currently runs simultaneously with StackRox Scanner
+
+1. Upgrades to the first installation will also upgrade the CRD used by any other installation, introducing the potential for versioning mismatches.
+2. Uninstallation of the first installation will remove the CRD, thus removing all custom `SecurityPolicy` resources from ALL Central installations.
+3. Downdgrading of the first installation could remove the CRD as well.
```
```suggestion
3. Downgrading of the first installation could remove the CRD as well.
```

porridge reviewed (changes requested):
I think this also deserves a rather prominent mention in the changelog, as users who used to be able to install multiple centrals using helm will now face an error and will need to find the documentation on how to deal with it.

Also, it bothers me a little that the same text is in two places. Maybe we could replace the bulk of one with a reference to the other file? 🤔 

And most importantly, this seems to break upgrades, the scanner-v4-install tests detected this 🎉 🥳 cc @mclasmeier 

I remember you had some instructions with annotation commands, I guess these need to be added and also prominently mentioned in the release notes as every (?) central helm chart user will need to run them before upgrade? 😟

kylape commented:
> Maybe we could replace the bulk of one with a reference to the other file?

I deleted most of the text from the values.yaml comment and added a reference to the readme.  I intentionally omitted the actual value from values.yaml to force users to go to the readme to get it.

> instructions with annotation commands [...] need to be added and also prominently mentioned in the release notes as every (?) central helm chart user will need to run them before upgrade?

Yes, that's correct.  I've added them to the changelog.

> Or should we add a pre-upgrade Job to apply the label and annotations to make this transparent?

What image could I use to run the job?  AFAICT, all the stackrox images don't have curl, wget, or kubectl available, so I'm not sure how I could run these commands.  Building a new binary to be included in the main image just for this one thing seems overkill, but that _would_ solve it.

clickboo commented on CHANGELOG.md:
```diff
@@ -31,6 +33,16 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
+  To avoid installation failure, Helm users need to apply the following changes to the CRD prior to upgrade:
+
+      kubectl annotate crd/securitypolicies.config.stackrox.io meta.helm.sh/release-name=stackrox-central-services
+      kubectl annotate crd/securitypolicies.config.stackrox.io meta.helm.sh/release-namespace=stackrox
```
We do allow configuring the namespace to other than stackrox right? So this line will have to accommodate that.

kylape commented:
/test ocp-4-12-nongroovy-e2e-tests

porridge commented on CHANGELOG.md:
```diff
@@ -22,6 +24,16 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
+- ROX-27622: Move `SecurityPolicy` CRD to template directory in Helm chart. **All Helm users will need to take action!**
+  No action is needed for users that use the operator or `roxctl` to install StackRox.
+  This change makes the CRD simpler to maintain for users because it will now be automatically upgraded.
+  To avoid installation failure, Helm users need to apply the following changes to the CRD prior to upgrade:
```
```suggestion
  To avoid upgrade failure, Helm users need to apply the following changes to the CRD prior to upgrade:
```

porridge commented on CHANGELOG.md:
```diff
@@ -22,6 +24,16 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
+      kubectl annotate crd/securitypolicies.config.stackrox.io meta.helm.sh/release-namespace=stackrox
+      kubectl label crd/securitypolicies.config.stackrox.io app.kubernetes.io/managed-by=Helm
+
+  The value of the annotations will need to be updated if you've updated the release name (i.e. "stackrox-central-services") or namespace (i.e. "stackrox").
```
This sentence could be misunderstood, given that the commands update annotation values. How about:

```suggestion
  The above values will need to be updated to match your release name (i.e. "stackrox-central-services") or namespace (i.e. "stackrox") in case you had used different ones.
```

porridge commented on tests/e2e/run-scanner-v4-install.bats:
```diff
@@ -253,6 +253,13 @@ setup() {
 }
 
+# CRD needs to be owned by Helm if upgrading to 4.7+ from 4.6.x via Helm
+apply_crd_ownership() {
```
Currently this is only invoked in one case where `stackrox` is the namespace, but please make it take the namespace as a parameter. Also a more specific name would be nice, how about:
```suggestion
apply_config_crd_ownership_for_upgrade() {
```

porridge reviewed (approved):
LGTM modulo the nitpicks inline but please let's have at least one of @mclasmeier or @vladbologa review as well.
Also please:
- write a KCS, I'm afraid this will cause multiple support tickets
- mention in PR description that you tried to automate this using a helm pre-upgrade hook but it's too late to help

kylape commented:
KCS draft: https://access.redhat.com/solutions/7114942

mclasmeier commented on CHANGELOG.md:
```diff
@@ -9,6 +9,8 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
 
 ## [NEXT RELEASE]
 
+**HELM USERS**: Please see the ROX-27622 under "technical changes" to avoid upgrade failures!
```
```suggestion
**HELM USERS**: Please see ROX-27622 under "technical changes" to avoid upgrade failures!
```
Not a native speaker, feel free to ignore.

mclasmeier commented on image/templates/helm/stackrox-central/internal/config-shape.yaml:
```diff
@@ -183,6 +183,7 @@ customize:
   other: {} # dict
 allowNonstandardNamespace: null # bool
 allowNonstandardReleaseName: null # bool
+allowUnsupportedMultipleCentralsUsingHelm: null # string
```
I was wondering if we need the "UsingHelm" suffix, since this describes a configuration value for the Helm chart. Feels somewhat unnecessary to me, but no strong feelings from my side.

kylape replied:
My read on the `UsingHelm` phrase is that it indicates that this feature (multiple centrals) is unsupported _specificially using Helm_.  We do support multiple Centrals using the operator.  That said, I think that bit of meaning is not particularly valuable, and the name is already long enough without the `UsingHelm` suffix.

kylape replied:
I'm adding a note to the helm readme:

> Note that installing multiple Centrals is fully supported by default using the ACS operator.

mclasmeier commented on image/templates/helm/stackrox-central/templates/00-securitypolicy-crd.yaml:
```diff
@@ -0,0 +1,4 @@
+{{- include "srox.init" . -}}
+{{ if ne (._rox.allowUnsupportedMultipleCentralsUsingHelm | toString) "documentationWasReadAndUnderstood" -}}
+{{ $.Files.Get "internal/config.stackrox.io_securitypolicies.yaml" }}
+{{- end -}}
```
I am advocating for keeping the templating style (especially with respect to whitespace elimination) somewhat consistent in our code base, hence I would like to suggest to follow [our current guide lines](https://github.com/stackrox/stackrox/blob/6ee9603c57f97efa33f272ed01815709b073d6dd/image/templates/CHANGING_CHARTS.md#L5).

```suggestion
{{- if ne (._rox.allowUnsupportedMultipleCentralsUsingHelm | toString) "documentationWasReadAndUnderstood" }}
{{ $.Files.Get "internal/config.stackrox.io_securitypolicies.yaml" }}
{{- end }}
```

mclasmeier commented on image/templates/helm/stackrox-central/templates/_init.tpl.htpl:
```diff
@@ -62,7 +62,7 @@
     {{ include "srox.note" (list $ (printf "You have chosen to deploy to namespace '%s'." $.Release.Namespace)) }}
   {{ else }}
-    {{ include "srox.fail" (printf "You have chosen to deploy to namespace '%s', not 'stackrox'. If this was accidental, please re-run helm with the '-n stackrox' option. Otherwise, if you need to deploy into this namespace, set the 'allowNonstandardNamespace' configuration value to true." $.Release.Namespace) }}
+    {{ include "srox.fail" (printf "You have chosen to deploy to namespace '%s', not 'stackrox'. If this was accidental, please re-run helm with the '-n stackrox' option. Otherwise, if you need to deploy into this namespace, set the 'allowNonstandardNamespace' configuration value to true. NOTE: Please review the section of the readme titled 'Multiple Installations of Central' if you encounter ins ...
```
```suggestion
    {{ include "srox.fail" (printf "You have chosen to deploy to namespace '%s', not 'stackrox'. If this was accidental, please re-run helm with the '-n stackrox' option. Otherwise, if you need to deploy into this namespace, set the 'allowNonstandardNamespace' configuration value to true. NOTE: Please review the section of this chart's README titled 'Multiple Installations of Central' if you encounter installation errors regarding a CustomResourceDefinition." $.Release.Namespace) }}
```

mclasmeier commented on image/templates/helm/stackrox-central/values.yaml.htpl:
```diff
@@ -552,6 +552,14 @@ central:
 ## default name, and make it a little harder to deviate from that.
 #allowNonstandardReleaseName: false
 #
+## This option is meant to be a very explicit acknowledgement by a user that that they understand
```
```suggestion
## This option is meant to be a very explicit acknowledgement by a user that they understand
```

mclasmeier commented on image/templates/helm/stackrox-central/values.yaml.htpl:
```diff
@@ -552,6 +552,14 @@ central:
+## the implications of installing more than one instance of Central on a single cluster, particularly
+## around CRD management.
+##
+## To enable the option, please review the readme section titled "Multiple Installations of Central"
```
```suggestion
## To enable the option, please review the README section titled "Multiple Installations of Central"
```

mclasmeier commented on tests/e2e/run-scanner-v4-install.bats:
```diff
@@ -257,6 +257,14 @@ setup() {
     _end
 }
 
+# CRD needs to be owned by Helm if upgrading to 4.7+ from 4.6.x via Helm
```
```suggestion
# CRD needs to be owned by Helm if upgrading to 4.8+ from 4.7.x via Helm
```

mclasmeier reviewed (approved):
Just a couple of nitpicks and smaller cosmetic suggestions. Thank you!
